### PR TITLE
feat(deploy-config): adds new abap deploy config writer

### PR DESCRIPTION
**Pull request by cianmSAP**
Adds new abap deploy config writer

### CODE REVIEW

**Comment by tobiasqueck** [comment]
@cianmSAP could you please also enhance the `create` module and use the write for the command `add deploy-config`. Then we have everything in the PR that allows e2e tests.

@testojs please verify that this also works for ADP (I don't see a reason why not but please check)

File: packages/abap-deploy-config-writer/src/config.ts [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    if (config.target.client) {
+        target.client = config.target.client;
+    }
+    if (config.target.scp) {
+        target.scp = true;
+    }
+
+    if (!isAppStudio()) {
+        const middleware = baseConfig.findCustomMiddleware<FioriToolsProxyConfig>('fiori-tools-proxy');
+        if (middleware?.configuration?.backend?.[0].authenticationType === 'reentranceTicket') {
```
Comment (devinea) [inline packages/abap-deploy-config-writer/src/config.ts:51]: accessing via '[0]' is not very robust. Maybe this should iterate over each backend and check for authenticationType

File: packages/abap-deploy-config-writer/src/index.ts [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    await updatePackageScript(basePath, 'deploy', deployScript, fs);
+    // undeploy script
+    const undeployScript = `${BUILD_SCRIPT} && ${UNDEPLOY_SCRIPT} --config ${deployConfigFile}`;
+    await updatePackageScript(basePath, 'undeploy', undeployScript, fs);
+    // test mode script
+    const deployTestModeScript = `${BUILD_SCRIPT} && ${DEPLOY_SCRIPT} --config ${deployConfigFile} --testMode true`;
+    await updatePackageScript(basePath, 'deploy-test', deployTestModeScript, fs);
+    // dependencies
+    await addPackageDevDependency(basePath, RIMRAF, RIMRAF_VERSION, fs);
+    await addPackageDevDependency(basePath, UI5_TASK_FLATTEN_LIB, UI5_TASK_FLATTEN_LIB_VERSION, fs);
```
Comment (devinea) [inline packages/abap-deploy-config-writer/src/index.ts:66]: Is this only for libs or also for ui5 applications?

File: packages/abap-deploy-config-writer/test/sample/test-ts-app/tsconfig.json [hunk 3]
```diff
@@ -0,0 +1,6 @@
+{
+    "compilerOptions": {
+        "target": "es2022"
```
Comment (devinea) [inline packages/abap-deploy-config-writer/test/sample/test-ts-app/tsconfig.json:3]:
```suggestion
        "target": "ES2022"
```

File: packages/abap-deploy-config-writer/src/index.ts [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+    addUi5Dependency(fs, basePath, UI5_TASK_FLATTEN_LIB);
```
Comment (devinea) [inline packages/abap-deploy-config-writer/src/index.ts:67]: addUi5Dependency is only valid for @ui5/cli less than v3.
https://sap.github.io/ui5-tooling/v3/updates/migrate-v3/#changes-to-dependency-configuration

**Review by devinea — COMMENTED** [review]
Looks good overall. 
New module setup ✅ 
project access  ✅ 
ui5-config ✅ 
yaml ✅ 
changeset  ✅

File: packages/abap-deploy-config-writer/src/index.ts (hunk 2 above)
Comment (cianmSAP) [reply packages/abap-deploy-config-writer/src/index.ts:66]: Currently it is added for both, and has been that way for a couple of years. But I can't see why it is added for ui5 applications. I'll add a condition on this

File: packages/abap-deploy-config-writer/src/index.ts (hunk 4 above)
Comment (cianmSAP) [reply packages/abap-deploy-config-writer/src/index.ts:67]: Ok I'll a check for the `@ui5/cli` to support adding deploy config to older apps

**Review by devinea — APPROVED** [review]
@cianmSAP Thanks for the updates. 
All good from my perspective.

File: packages/project-access/src/project/script.ts [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+
+/**
+ * Updates the package.json with a new script.
+ *
+ * @param basePath - the base path
+ * @param scriptName - the script name
+ * @param script - the script content
+ * @param fs - optional memfs editor instance
+ */
+export async function updatePackageScript(
```
Comment (longieirl) [inline packages/project-access/src/project/script.ts:15]: Thanks for this! I'll need this in the CF writer.

**Pull request merged**